# Session 2: Essential Pandas for Data Science

<hr>

### Agenda
1. Introduction to Pandas
2. Understanding Series & DataFrames
3. Loading CSV,JSON
4. Connecting databases
5. Descriptive Statistics
6. Accessing subsets of data - Rows, Columns, Filters
7. Handling Missing Data
8. Dropping rows & columns
9. Handling Duplicates
10. Function Application - map, apply, groupby, rolling, str
11. Merge, Join & Concatenate
12. Pivot-tables
13. Normalizing JSON

<hr>

### 1. Introduction to Pandas
* High Performance, Easy-to-use open source library for Data Analysis
* Creates tabular format of data from different sources like csv, json, database.
* Have utilities for descriptive statistics, aggregation, handling missing data
* Database utilities like merge, join are available
* Fast, Programmable & Easy alternative to spreadsheets

In [19]:
import pandas as pd
import numpy as np

### 2. Understanding Series & DataFrames
* Series represents one column
* Combine multiple columns to create a table ( .i.e DataFrame )

In [20]:
ser1 = pd.Series(data=[1,2,3,4,5], index=list('abcde'))
ser1

a    1
b    2
c    3
d    4
e    5
dtype: int64

In [21]:
ser2 = pd.Series(data=[11,22,33,44,55], index=list('abcde'))
ser2

a    11
b    22
c    33
d    44
e    55
dtype: int64

* Creating DataFrame from above two series
* Data corresponding to same index belongs to same row

In [22]:
df = pd.DataFrame({'A':ser1, 'B':ser2})
df

A   B
a  1  11
b  2  22
c  3  33
d  4  44
e  5  55

* Creating a random dataframe of 10 X 10

In [23]:
pd.DataFrame(data=np.random.randint(1,10,size=(10,10)), index=list('ABCDEFGHIJ'), columns=list('abcdefghij'))

a  b  c  d  e  f  g  h  i  j
A  9  4  3  7  8  4  9  1  9  7
B  9  1  5  4  2  1  9  3  9  4
C  8  3  3  7  1  3  9  8  6  7
D  9  5  9  4  1  1  6  7  3  9
E  8  4  2  4  3  5  9  7  2  4
F  9  8  5  5  1  9  4  5  6  1
G  2  4  1  3  4  5  4  2  6  8
H  3  8  3  1  9  3  2  3  8  3
I  4  3  3  9  1  9  6  6  4  1
J  2  3  1  2  8  9  3  5  7  1

### 3. Loading CSV,JSON

In [24]:
hr_data = pd.read_csv('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/HR_comma_sep.csv.txt')

In [25]:
hr_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   sales                  14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


In [26]:
hr_data.to_numpy()

array([[0.38, 0.53, 2, ..., 0, 'sales', 'low'],
       [0.8, 0.86, 5, ..., 0, 'sales', 'medium'],
       [0.11, 0.88, 7, ..., 0, 'sales', 'medium'],
       ...,
       [0.37, 0.53, 2, ..., 0, 'support', 'low'],
       [0.11, 0.96, 6, ..., 0, 'support', 'low'],
       [0.37, 0.52, 2, ..., 0, 'support', 'low']], dtype=object)

In [27]:
hr_data_itr = pd.read_csv('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/HR_comma_sep.csv.txt', chunksize=5000)

In [28]:
for hr_data in hr_data_itr:
    print (hr_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     5000 non-null   float64
 1   last_evaluation        5000 non-null   float64
 2   number_project         5000 non-null   int64  
 3   average_montly_hours   5000 non-null   int64  
 4   time_spend_company     5000 non-null   int64  
 5   Work_accident          5000 non-null   int64  
 6   left                   5000 non-null   int64  
 7   promotion_last_5years  5000 non-null   int64  
 8   sales                  5000 non-null   object 
 9   salary                 5000 non-null   object 
dtypes: float64(2), int64(6), object(2)
memory usage: 390.8+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 5000 to 9999
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 ----------

In [29]:
pd.read_json('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/movie.json.txt')

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Vertigo                4.0              3.0         4.5            NaN   
Scarface               4.5              1.5         5.0            3.5   
Raging Bull            3.0              1.0         NaN            5.0   
Goodfellas             4.5              2.0         4.5            5.0   
The Apartment          1.0              5.0         1.0            1.0   
Roman Holiday          NaN              4.5         NaN            1.0   

               Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Vertigo                 NaN               5.0         3.5           NaN  
Scarface                2.5               4.5         3.0           NaN  
Raging Bull             NaN               4.0         NaN           4.5  
Goodfellas              3.0               2.5         4.5           NaN  
The Apartment           NaN               1.0         1.0           1.5  
Roman Holiday           4.5               1.5         3.0           NaN

### 4. Connecting Databases

In [30]:
!pip install pysqlite3

In [31]:
# import sqlite3
# con = sqlite3.connect('Data/database.sqlite')
# pd.read_sql_query("SELECT * FROM Reviews LIMIT 5",con)

* import MySQLdb
* mysql_cn= MySQLdb.connect(host='myhost', 
                port=3306,user='myusername', passwd='mypassword', 
                db='information_schema')
* df_mysql = pd.read_sql('select * from VIEWS;', con=mysql_cn)

### 5. Descriptive Statistics
* Pandas api's for understanding data

In [32]:
hr_data = pd.read_csv('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/HR_comma_sep.csv.txt')

In [33]:
hr_data.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years  sales  \
0                   3              0     1                      0  sales   
1                   6              0     1                      0  sales   
2                   4              0     1                      0  sales   
3                   5              0     1                      0  sales   
4                   3              0     1                      0  sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [34]:
hr_data.tail()

satisfaction_level  last_evaluation  number_project  \
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years    sales salary  
14994                      0  support    low  
14995                      0  support    low  
14996                      0  support    low  
14997                      0  support    low  
14998                      0  support    low

In [35]:
hr_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   sales                  14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


In [36]:
hr_data.describe()

satisfaction_level  last_evaluation  number_project  \
count        14999.000000     14999.000000    14999.000000   
mean             0.612834         0.716102        3.803054   
std              0.248631         0.171169        1.232592   
min              0.090000         0.360000        2.000000   
25%              0.440000         0.560000        3.000000   
50%              0.640000         0.720000        4.000000   
75%              0.820000         0.870000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident          left  \
count          14999.000000        14999.000000   14999.000000  14999.000000   
mean             201.050337            3.498233       0.144610      0.238083   
std               49.943099            1.460136       0.351719      0.425924   
min               96.000000            2.000000       0.000000      0.000000   
25%              156.000000            3.000000       0.000000      0.000000   
50%              200.000000            3.000000       0.000000      0.000000   
75%              245.000000            4.000000       0.000000      0.000000   
max              310.000000           10.000000       1.000000      1.000000   

       promotion_last_5years  
count           14999.000000  
mean                0.021268  
std                 0.144281  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 0.000000  
max                 1.000000

In [37]:
hr_data.salary.value_counts()

low       7316
medium    6446
high      1237
Name: salary, dtype: int64

### 6. Accessing subset of data - rows, columns, filters
* Get all columns with categorical values

In [38]:
cat_cols_data = hr_data.select_dtypes('object')

In [39]:
cat_cols_data.head()

sales  salary
0  sales     low
1  sales  medium
2  sales  medium
3  sales     low
4  sales     low

* Rename columns names

In [40]:
hr_data.rename(columns={'sales':'department'},inplace=True)

In [41]:
hr_data.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

* Select column by column names

In [42]:
hr_data.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'department', 'salary'],
      dtype='object')

In [43]:
hr_data[['satisfaction_level','last_evaluation','number_project']].head()

satisfaction_level  last_evaluation  number_project
0                0.38             0.53               2
1                0.80             0.86               5
2                0.11             0.88               7
3                0.72             0.87               5
4                0.37             0.52               2

In [44]:
hr_data.satisfaction_level[:5]

0    0.38
1    0.80
2    0.11
3    0.72
4    0.37
Name: satisfaction_level, dtype: float64

In [45]:
hr_data['satisfaction_level'][:5]

0    0.38
1    0.80
2    0.11
3    0.72
4    0.37
Name: satisfaction_level, dtype: float64

In [46]:
movie_data = pd.read_json('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/movie.json.txt')

In [47]:
movie_data

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Vertigo                4.0              3.0         4.5            NaN   
Scarface               4.5              1.5         5.0            3.5   
Raging Bull            3.0              1.0         NaN            5.0   
Goodfellas             4.5              2.0         4.5            5.0   
The Apartment          1.0              5.0         1.0            1.0   
Roman Holiday          NaN              4.5         NaN            1.0   

               Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Vertigo                 NaN               5.0         3.5           NaN  
Scarface                2.5               4.5         3.0           NaN  
Raging Bull             NaN               4.0         NaN           4.5  
Goodfellas              3.0               2.5         4.5           NaN  
The Apartment           NaN               1.0         1.0           1.5  
Roman Holiday           4.5               1.5         3.0           NaN

* Access data by index values

In [48]:
movie_data.loc['Scarface']

David Smith         4.5
Brenda Peterson     1.5
Bill Duffy          5.0
Samuel Miller       3.5
Julie Hammel        2.5
Clarissa Jackson    4.5
Adam Cohen          3.0
Chris Duncan        NaN
Name: Scarface, dtype: float64

In [49]:
movie_data.loc['Scarface':'Vertigo']

Empty DataFrame
Columns: [David Smith, Brenda Peterson, Bill Duffy, Samuel Miller, Julie Hammel, Clarissa Jackson, Adam Cohen, Chris Duncan]
Index: []

In [50]:
movie_data['Scarface':'Vertigo']

Empty DataFrame
Columns: [David Smith, Brenda Peterson, Bill Duffy, Samuel Miller, Julie Hammel, Clarissa Jackson, Adam Cohen, Chris Duncan]
Index: []

In [51]:
movie_data.iloc[1]

David Smith         4.5
Brenda Peterson     1.5
Bill Duffy          5.0
Samuel Miller       3.5
Julie Hammel        2.5
Clarissa Jackson    4.5
Adam Cohen          3.0
Chris Duncan        NaN
Name: Scarface, dtype: float64

In [52]:
movie_data.iloc[1:4]

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Scarface             4.5              1.5         5.0            3.5   
Raging Bull          3.0              1.0         NaN            5.0   
Goodfellas           4.5              2.0         4.5            5.0   

             Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Scarface              2.5               4.5         3.0           NaN  
Raging Bull           NaN               4.0         NaN           4.5  
Goodfellas            3.0               2.5         4.5           NaN

In [53]:
movie_data[1:4]

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Scarface             4.5              1.5         5.0            3.5   
Raging Bull          3.0              1.0         NaN            5.0   
Goodfellas           4.5              2.0         4.5            5.0   

             Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Scarface              2.5               4.5         3.0           NaN  
Raging Bull           NaN               4.0         NaN           4.5  
Goodfellas            3.0               2.5         4.5           NaN

* Filtering rows based on conditions

In [54]:
movie_data[ (movie_data['Adam Cohen'] > 3)]

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Vertigo             4.0              3.0         4.5            NaN   
Goodfellas          4.5              2.0         4.5            5.0   

            Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Vertigo              NaN               5.0         3.5           NaN  
Goodfellas           3.0               2.5         4.5           NaN

In [55]:
movie_data[ ((movie_data['Adam Cohen'] > 3) & (movie_data['David Smith'] > 4))]

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Goodfellas          4.5              2.0         4.5            5.0   

            Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Goodfellas           3.0               2.5         4.5           NaN

### 7. Handling missing data
* Machine Learning algorithms don't expect data missing
* If there is a columns with more than 40% data missing, we may drop the column
* Fow rows with, important column values missing. Drop the rows

In [56]:
movie_data

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Vertigo                4.0              3.0         4.5            NaN   
Scarface               4.5              1.5         5.0            3.5   
Raging Bull            3.0              1.0         NaN            5.0   
Goodfellas             4.5              2.0         4.5            5.0   
The Apartment          1.0              5.0         1.0            1.0   
Roman Holiday          NaN              4.5         NaN            1.0   

               Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Vertigo                 NaN               5.0         3.5           NaN  
Scarface                2.5               4.5         3.0           NaN  
Raging Bull             NaN               4.0         NaN           4.5  
Goodfellas              3.0               2.5         4.5           NaN  
The Apartment           NaN               1.0         1.0           1.5  
Roman Holiday           4.5               1.5         3.0           NaN

* Get all the rows for which column 'Bill Duffy' is missing

In [57]:
movie_data['Bill Duffy'].notnull()

Vertigo           True
Scarface          True
Raging Bull      False
Goodfellas        True
The Apartment     True
Roman Holiday    False
Name: Bill Duffy, dtype: bool

In [58]:
movie_data[movie_data['Bill Duffy'].notnull()]

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Vertigo                4.0              3.0         4.5            NaN   
Scarface               4.5              1.5         5.0            3.5   
Goodfellas             4.5              2.0         4.5            5.0   
The Apartment          1.0              5.0         1.0            1.0   

               Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Vertigo                 NaN               5.0         3.5           NaN  
Scarface                2.5               4.5         3.0           NaN  
Goodfellas              3.0               2.5         4.5           NaN  
The Apartment           NaN               1.0         1.0           1.5

* Get all the rows for which 'Bill Duffy' is null

In [59]:
movie_data[movie_data['Bill Duffy'].isnull()]

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Raging Bull            3.0              1.0         NaN            5.0   
Roman Holiday          NaN              4.5         NaN            1.0   

               Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Raging Bull             NaN               4.0         NaN           4.5  
Roman Holiday           4.5               1.5         3.0           NaN

### 8. Dropping Rows & Columns

In [60]:
titanic_data = pd.read_csv('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/titanic-train.csv.txt')

In [61]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


* Dropping 'Cabin' column as it has only 204 data present in 891 rows

In [62]:
titanic_data.drop(['Cabin'],axis=1,inplace=True)

In [63]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


* Now, drop all rows with missing values
* We don't have inplace = True, so doesn't modify the dataframe

In [64]:
titanic_data.dropna().info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 712 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  712 non-null    int64  
 1   Survived     712 non-null    int64  
 2   Pclass       712 non-null    int64  
 3   Name         712 non-null    object 
 4   Sex          712 non-null    object 
 5   Age          712 non-null    float64
 6   SibSp        712 non-null    int64  
 7   Parch        712 non-null    int64  
 8   Ticket       712 non-null    object 
 9   Fare         712 non-null    float64
 10  Embarked     712 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 66.8+ KB


* Consider only selected columns to check if they contain NA

In [65]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


In [66]:
titanic_data.dropna(subset=['Embarked','Age']).info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 712 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  712 non-null    int64  
 1   Survived     712 non-null    int64  
 2   Pclass       712 non-null    int64  
 3   Name         712 non-null    object 
 4   Sex          712 non-null    object 
 5   Age          712 non-null    float64
 6   SibSp        712 non-null    int64  
 7   Parch        712 non-null    int64  
 8   Ticket       712 non-null    object 
 9   Fare         712 non-null    float64
 10  Embarked     712 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 66.8+ KB


* Another approach of handling missing data is filling the missing ones

In [67]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


In [68]:
titanic_data.fillna({'Age':0,'Embarked':'Unknown'}).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


In [69]:
titanic_data.Age.fillna(method='ffill')[:5]
#Other options are 'bfill'

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

### 9. Handling Duplicates
* Sometimes, it difficult to ensure that data is not duplicated.
* This becomes responsibility in Data cleaning step to make sure duplicated data is deleted

In [70]:
df = pd.DataFrame({'A':[1,1,3,4,5,1], 'B':[1,1,3,7,8,1], 'C':[3,1,1,6,7,1]})

In [71]:
df

A  B  C
0  1  1  3
1  1  1  1
2  3  3  1
3  4  7  6
4  5  8  7
5  1  1  1

In [72]:
df.duplicated()

0    False
1    False
2    False
3    False
4    False
5     True
dtype: bool

In [73]:
df[df.duplicated()]

A  B  C
5  1  1  1

In [74]:
df[df.duplicated(subset=['A','B'])]

A  B  C
1  1  1  1
5  1  1  1

### 10. Function Application
* map for transforming one column to another
* Can be applied only to series

In [75]:
titanic_data_age = titanic_data[titanic_data.Age.notnull()]

In [76]:
titanic_data['age_category'] = titanic_data.Age.map(lambda age: 'Kid' if age < 18 else 'Adult')

In [77]:
titanic_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked age_category  
0      0         A/5 21171   7.2500        S        Adult  
1      0          PC 17599  71.2833        C        Adult  
2      0  STON/O2. 3101282   7.9250        S        Adult  
3      0            113803  53.1000        S        Adult  
4      0            373450   8.0500        S        Adult

* apply function can be done to Series as well as DataFrames

In [78]:
titanic_data.Age.apply('sum')

21205.17

In [79]:
titanic_data.Age.apply(lambda age: 'Kid' if age < 18 else 'Adult')[:10]

0    Adult
1    Adult
2    Adult
3    Adult
4    Adult
5    Adult
6    Adult
7      Kid
8    Adult
9      Kid
Name: Age, dtype: object

* apply on dataframes helps us dealing with multiple columns
* func will receive all the rows

In [80]:
#e will be each row
def func(e):
    if e.Sex == 'male':
        return e.Fare * 2
    else:
        return e.Fare
    

In [81]:
titanic_data.apply(func,axis=1)[:5]

0    14.5000
1    71.2833
2     7.9250
3    53.1000
4    16.1000
dtype: float64

* groupby - It splits data into groups, a function is applied to each groups separately, combine results into a data structure

In [82]:
titanic_data.groupby(['Sex']).Age.mean()

Sex
female    27.915709
male      30.726645
Name: Age, dtype: float64

In [83]:
titanic_data.groupby(['Sex']).Age.agg(['mean','min','max'])

mean   min   max
Sex                          
female  27.915709  0.75  63.0
male    30.726645  0.42  80.0

* Rolling for window based operation

In [84]:
titanic_data.Age.rolling(window=5,min_periods=1).agg(['sum','min'])

sum   min
0     22.0  22.0
1     60.0  22.0
2     86.0  22.0
3    121.0  22.0
4    156.0  22.0
..     ...   ...
886  141.0  22.0
887  138.0  19.0
888  110.0  19.0
889  111.0  19.0
890  104.0  19.0

[891 rows x 2 columns]

* For columns containing string, we have str utilities

In [85]:
titanic_data[titanic_data.Name.str.contains('Mr')]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
3              4         1       1   
4              5         0       3   
5              6         0       3   
..           ...       ...     ...   
883          884         0       2   
884          885         0       3   
885          886         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
..                                                 ...     ...   ...    ...   
883                      Banfield, Mr. Frederick James    male  28.0      0   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked age_category  
0        0         A/5 21171   7.2500        S        Adult  
1        0          PC 17599  71.2833        C        Adult  
3        0            113803  53.1000        S        Adult  
4        0            373450   8.0500        S        Adult  
5        0            330877   8.4583        Q        Adult  
..     ...               ...      ...      ...          ...  
883      0  C.A./SOTON 34068  10.5000        S        Adult  
884      0   SOTON/OQ 392076   7.0500        S        Adult  
885      5            382652  29.1250        Q        Adult  
889      0            111369  30.0000        C        Adult  
890      0            370376   7.7500        Q        Adult  

[647 rows x 12 columns]

### 11. Append,Merge, Join & Concatenate
* Append for stacking dataframe

In [86]:
df1 = pd.DataFrame(data=np.random.randint(1,10,size=(10,3)), columns=list('ABC'))

In [87]:
df2 = pd.DataFrame(data=np.random.randint(1,10,size=(10,3)), columns=list('ABC'))

In [88]:
df1

A  B  C
0  6  9  1
1  4  3  2
2  2  1  5
3  2  9  8
4  5  9  5
5  2  5  2
6  1  4  1
7  8  9  6
8  5  4  1
9  2  2  4

In [89]:
df2

A  B  C
0  7  9  1
1  5  7  4
2  7  8  2
3  2  6  8
4  1  1  8
5  3  9  4
6  3  5  7
7  5  5  6
8  7  3  4
9  4  9  8

In [90]:
df1.append(df2, ignore_index=True)

A  B  C
0   6  9  1
1   4  3  2
2   2  1  5
3   2  9  8
4   5  9  5
5   2  5  2
6   1  4  1
7   8  9  6
8   5  4  1
9   2  2  4
10  7  9  1
11  5  7  4
12  7  8  2
13  2  6  8
14  1  1  8
15  3  9  4
16  3  5  7
17  5  5  6
18  7  3  4
19  4  9  8

In [91]:
left = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3','K4','K5'],
                         'A': ['A0', 'A1', 'A2', 'A3','A4','A5'],
                         'B': ['B0', 'B1', 'B2', 'B3','B4','B5']})

right = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3','K6','K7'],
                          'C': ['C0', 'C1', 'C2', 'C3','C6','C7'],
                         'D': ['D0', 'D1', 'D2', 'D3','D6','D7']})

In [92]:
left.merge(right, on='key')

key   A   B   C   D
0  K0  A0  B0  C0  D0
1  K1  A1  B1  C1  D1
2  K2  A2  B2  C2  D2
3  K3  A3  B3  C3  D3

In [93]:
left.merge(right, on='key', how='left')

key   A   B    C    D
0  K0  A0  B0   C0   D0
1  K1  A1  B1   C1   D1
2  K2  A2  B2   C2   D2
3  K3  A3  B3   C3   D3
4  K4  A4  B4  NaN  NaN
5  K5  A5  B5  NaN  NaN

* join for combining data based on index values

In [94]:
left = pd.DataFrame({'A': ['A0', 'A1', 'A2'],
                     'B': ['B0', 'B1', 'B2']},
                     index=['K0', 'K1', 'K2'])

right = pd.DataFrame({'C': ['C0', 'C2', 'C3'],
                      'D': ['D0', 'D2', 'D3']},
                     index=['K0', 'K2', 'K3'])

In [95]:
left.join(right)

A   B    C    D
K0  A0  B0   C0   D0
K1  A1  B1  NaN  NaN
K2  A2  B2   C2   D2

### 12. Pivot Tables
* An useful way to get important information from data

In [96]:
sales_data = pd.read_excel('https://github.com/zekelabs/data-science-complete-tutorial/blob/master/Data/sales-funnel.xlsx?raw=true')

ImportError: Missing optional dependency 'xlrd'. Install xlrd >= 1.0.0 for Excel support Use pip or conda to install xlrd.

In [ ]:
sales_data

In [ ]:
pd.pivot_table(sales_data, index=['Manager','Rep'], values=['Account','Price'], aggfunc=[np.sum, np.mean])

### 13. Normalizing JSON
* JSON data will not always be of flat but can be hierchial

In [ ]:
data = [     {'state': 'Florida',
              'shortname': 'FL',
              'info': {
                   'governor': 'Rick Scott'
              },
              'counties': [{'name': 'Dade', 'population': 12345},
                          {'name': 'Broward', 'population': 40000},
                          {'name': 'Palm Beach', 'population': 60000}]},
             {'state': 'Ohio',
              'shortname': 'OH',
              'info': {
                   'governor': 'John Kasich'
              },
              'counties': [{'name': 'Summit', 'population': 1234},
                           {'name': 'Cuyahoga', 'population': 1337}]}]

In [ ]:
from pandas.io.json import json_normalize

In [ ]:
json_normalize(data)

In [ ]:
json_normalize(data,'counties',['state',['info', 'governor']])